### PR TITLE
Split Federation Services and Gateway content into separate articles

### PR DESCRIPTION
**Pull request by StephenBarlow**

### CODE REVIEW

File: docs/source/federation/gateway.mdx [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+  // highlight-end
+});
+
+const server = new ApolloServer({
+  gateway,
+
+  // Disable subscriptions (not currently supported with ApolloGateway)
+  subscriptions: false,
+
+  context: ({ req }) => {
```
Comment (abernix) [inline docs/source/federation/gateway.mdx]: I know this focuses on `apollo-server`, which uses `apollo-server-express` (and eventually, `express`) under the the hood, but in the same spirit as my comment on https://github.com/apollographql/apollo-server/pull/3834/files#r391677331, it may be worth a non-normative note that specifies that this will be different on non-`express` integrations (e.g. Fastify, Koa, AWS Lambda, Google Cloud Functions).

The risk of course is copy-pasta and subtly unexpected behavior that results from such, but it's a state we've seen users encounter a few times and is slated to be improved in the implementation, _eventually_.

File: docs/source/federation/gateway.mdx [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+  loop For each operation in the query plan
+    Gateway->>Implementing Services: Sends the operation to the applicable service
+    Implementing Services->>Gateway: Responds with result and Server-Id header
+    Gateway->>Gateway: Adds the returned Server-Id to the shared context
+  end
+  Gateway->>Gateway: Adds all Server-Ids in the shared context to the response header
+  Gateway->>Client app: Sends operation response
+```
+
+To implement this flow, we can use the `didReceiveResponse` callback of the `RemoteGraphQLDataSource` class to inspect each implementing service's result as it comes in. We can add the `Server-Id` to the shared `context` in this callback, then pull the full list from the `context` when sending the final response to the client.
```
Comment (abernix) [inline docs/source/federation/gateway.mdx]: I love the way this so eloquently explains what the shape of the solution is _in words_ and the solution (_in code_) is so clearly demonstrated within the expando-example.

File: docs/source/federation/gateway.mdx [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+const { ApolloServer } = require('apollo-server');
+const { ApolloGateway, RemoteGraphQLDataSource } = require('@apollo/gateway');
+
+class DataSourceWithServerId extends RemoteGraphQLDataSource {
+  // highlight-start
+  async didReceiveResponse({ response, request, context }) {
+    // Parse the Server-Id header and add it to the array on context
+    const serverId = response.http.headers.get('Server-Id');
+    if (serverId) {
+      context.serverIds.push(serverId);
```
Comment (abernix) [inline docs/source/federation/gateway.mdx]: You've taken care in your implementation so this isn't a problem, though I feel a warning may be in order — similar to the warning which we (now?) give with regards to adjusting `context` within resolvers in the Resolvers page (after https://github.com/apollographql/apollo-server/pull/3834).  

In other words, guidance which advises to not make assumptions about the order of a services' `buildService` invocation, since that can change dynamically depending on the shape of the query and is determined by the query planner.

In other words, setting `context.prop` in one service is not necessarily guaranteed to happen before another service steps on it by doing the same thing.  Therefore, `push`-ing into, e.g., an Array, or an object keyed by the service's name is a more defensive approach (though while `buildService` receives the `name` of a service, the default `RemoteGraphQLDataSource` doesn't accept `name` itself. 🤔  Anyhow, I think my point stands and it could be keyed by something else as well).

File: docs/source/federation/gateway.mdx [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+    return { serverIds: [] };
+  },
+  plugins: [
+    // highlight-start
+    {
+      requestDidStart() {
+        return {
+          willSendResponse({ context, response }) {
+            // Append our final result to the outgoing response headers
+            response.http.headers.append(
```
Comment (abernix) [inline docs/source/federation/gateway.mdx]: It might be sometimes desirable, but it's somewhat surprising to me that the existing code didn't use `set` rather than [`append`](https://developer.mozilla.org/en-US/docs/Web/API/Headers/append) since it feels like this could create a condition with a comma-delimited list of comma-delimited attributes if `Server-Id` had already been set by another plugin or in other code?

e.g., ending up with potentially:

```
Server-Id: server1,server2 server1,server5
```

Not asking you for changes on this, but @trevor-scheer for input/thoughts.

File: docs/source/federation/gateway.mdx [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+
+## Implementing custom directives
+
+The `@apollo/gateway` library provides limited support for custom [directives](/schema/directives/) that are implemented by your implementing services. Directives must meet all of the following criteria to be compatible with the gateway:
+
+* Custom directives can only be used in **executable locations**, as defined in the [GraphQL specification](https://graphql.github.io/graphql-spec/June2018/#ExecutableDirectiveLocation).
+* Your custom directives must be defined by *every* implementing service in your data graph, even if not all of your services use them.
+* Directive definitions must be identical across all of your service schemas. Directive definitions are identical if their names, locations, arguments, and argument types are all identical.
+
+>Apollo Server does not currently support executable directives, however they are supported by the gateway.
```
Comment (abernix) [inline docs/source/federation/gateway.mdx]: This seems like it could be more prominent in this section — immediately below the heading, and even possibly just suggest that the intended target for this implementation suggestion is either users who have explicitly used `graphql-tools` tooling to augment a schema with directives or have built a schema in another way other than using Apollo Server's normal `typeDefs` behavior, or those who are using a completely different downstream server written in another language with Apollo Gateway.

File: docs/source/federation/gateway.mdx [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+
+To learn more about `buildService` and `RemoteGraphQLDataSource`, see the [API docs](/api/apollo-gateway/).
+
+## Implementing custom directives
+
+The `@apollo/gateway` library provides limited support for custom [directives](/schema/directives/) that are implemented by your implementing services. Directives must meet all of the following criteria to be compatible with the gateway:
+
+* Custom directives can only be used in **executable locations**, as defined in the [GraphQL specification](https://graphql.github.io/graphql-spec/June2018/#ExecutableDirectiveLocation).
+* Your custom directives must be defined by *every* implementing service in your data graph, even if not all of your services use them.
+* Directive definitions must be identical across all of your service schemas. Directive definitions are identical if their names, locations, arguments, and argument types are all identical.
```
Comment (abernix) [inline docs/source/federation/gateway.mdx]: @trevor-scheer Can you confirm the accuracy of these current claims of directive support?

File: docs/source/federation/gateway.mdx (hunk 4 above)
Comment (trevor-scheer) [reply docs/source/federation/gateway.mdx]: @abernix that's right, I don't see any reason why this shouldn't be `set` instead.

File: docs/source/federation/gateway.mdx (hunk 6 above)
Comment (trevor-scheer) [reply docs/source/federation/gateway.mdx]: @StephenBarlow this is no longer accurate, which is my fault. Happy to sync up with you if you'd like to build the understanding, else I can suggest some changes. Let me know what you'd prefer!

File: docs/source/federation/gateway.mdx (hunk 4 above)
Comment (abernix) [inline docs/source/federation/gateway.mdx]: Ok, asking for changes now, but not the PR author's fault. :wink:

```suggestion
            response.http.headers.set(
```

File: docs/source/federation/gateway.mdx [hunk 7]
```diff
@@ -216,10 +216,52 @@ To learn more about `buildService` and `RemoteGraphQLDataSource`, see the [API d
...
+query {
+  users {
+    name @uppercase
+  }
+}
+```
+
+At composition time, `ApolloGateway` makes sure that **all of your implementing services define the exact same set of executable directives**. If any service is missing a definition, or if definitions differ in their locations, arguments, or argument types, a composition error occurs.
+
+> Note that different implementing services _can_ use different logic to handle the same executable directive, but this is not recommended.
```
Comment (trevor-scheer) [inline docs/source/federation/gateway.mdx]: Can we flip this around and say something more like:
>It's strongly recommended that executable directives implement the same logic across different services. Failure to do so may result in a confusing and inconsistent experience for consumers of the graph.
